Objects/listobject.c

```diff
@@ -1901,6 +1918,186 @@ reverse_sortslice(sortslice *s, Py_ssize_t n)
+        ok = (res == Py_True);
+    }
+    else {
+        ok = PyObject_IsTrue(res);
+    }
+    Py_DECREF(res);
+    return ok;
+}
+
+/* Latin string compare: safe for any two latin (one byte per char) strings. */
+static int
+unsafe_latin_compare(PyObject* v, PyObject* w, MergeState* ms){
+  /* Modified from Objects/unicodeobject.c:unicode_compare, assuming: */
+  #ifdef Py_DEBUG
+    assert(v->ob_type == w->ob_type &&
+           v->ob_type == &PyUnicode_Type &&
+           PyUnicode_KIND(v) == PyUnicode_KIND(w) &&
+           PyUnicode_KIND(v) == PyUnicode_1BYTE_KIND);
+  #endif
+    
+    int len = Py_MIN(PyUnicode_GET_LENGTH(v), PyUnicode_GET_LENGTH(w));
+    int res = memcmp(PyUnicode_DATA(v), PyUnicode_DATA(w), len);
+
+    return (res != 0 ?
+            res < 0 :
+            PyUnicode_GET_LENGTH(v) < PyUnicode_GET_LENGTH(w));
+}
+
+/* Bounded int compare: compare any two longs that fit in a single machine word. */
+static int
+unsafe_long_compare(PyObject *v, PyObject *w, MergeState* ms)
+{
+  /* Modified from Objects/longobject.c:long_compare, assuming: */
+  #ifdef Py_DEBUG
+    assert(v->ob_type == w->ob_type &&
+           v->ob_type == &PyLong_Type &&
+           Py_ABS(Py_SIZE(v)) <= 1 &&
+           Py_ABS(Py_SIZE(w)) <= 1);
+  #endif
+    
+    PyLongObject *vl, *wl;
+    vl = (PyLongObject*)v;
+    wl = (PyLongObject*)w;
+
+    sdigit v0 = Py_SIZE(vl) == 0 ? 0 : (sdigit)vl->ob_digit[0];
+    sdigit w0 = Py_SIZE(wl) == 0 ? 0 : (sdigit)wl->ob_digit[0];
+
+    if (Py_SIZE(vl) < 0)
+        v0 = -v0;
+    if (Py_SIZE(wl) < 0)
+        w0 = -w0;
+
+    return v0 < w0;
+}
+
+/* Float compare: compare any two floats. */
+static int
+unsafe_float_compare(PyObject *v, PyObject *w, MergeState* ms){
+  /* Modified from Objects/floatobject.c:float_richcompare, assuming: */
+  #ifdef Py_DEBUG
+    assert(v->ob_type == w->ob_type &&
+           v->ob_type == &PyFloat_Type);
+  #endif
+    if (v == w) return 0;
+    
+    return PyFloat_AS_DOUBLE(v) < PyFloat_AS_DOUBLE(w);
+}
+
+/* Tuple compare: compare any two non-empty tuples, using 
+ * ms->tuple_elem_compare to compare the first elements, which is set 
+ * using the same pre-sort check as we use for ms->key_compare,
+ * but run on the list [x[0] for x in L]. This allows us to optimize compares
+ * on two levels (as long as [x[0] for x in L] is type-homogeneous.) */
+static int
+unsafe_tuple_compare(PyObject* v, PyObject* w, MergeState* ms)
+{
+  /* Modified from Objects/tupleobject.c:tuplerichcompare, assuming: */
+  #ifdef Py_DEBUG
+    assert(v->ob_type == w->ob_type &&
+           v->ob_type == &PyTuple_Type &&
+           Py_SIZE(v) > 0 &&
+           Py_SIZE(w) > 0);
+  #endif
+    
+    PyTupleObject *vt, *wt;
+    Py_ssize_t i;
+    Py_ssize_t vlen, wlen;
+
+    vt = (PyTupleObject *)v;
+    wt = (PyTupleObject *)w;
+    
+    int k;
+    
+    if (v == w) return 0;
+
+    /* Is v[0] < w[0]? */
+    k = (*(ms->tuple_elem_compare))(vt->ob_item[0], wt->ob_item[0], ms);
```

Comment:
No. If the compare function is `unsafe_object_compare`, then it could need the `ms` struct. For example: `[(2**60,), (-2**60,)]`. `unsafe_long_compare` doesn't apply because the ints are too big.